authorize/grpc.go

```diff
@@ -31,84 +41,87 @@ func (a *Authorize) Check(ctx context.Context, in *envoy_service_auth_v2.CheckRe
-			rawJWT = newRawJWT
-			sessionErr = nil
-			isNewSession = true
-		} else {
-			log.Warn().Err(err).Msg("authorize: error refreshing session")
-			// set the error to expired so that we can force a new login
-			sessionErr = sessions.ErrExpired
-		}
+	if sessionState != nil {
+		a.waitForSessionAndUser(time.Second*3, sessionState.ID)
 	}
 
-	req := getEvaluatorRequestFromCheckRequest(in, rawJWT)
-	reply, err := a.pe.IsAuthorized(ctx, req)
+	a.dataBrokerDataLock.RLock()
+	defer a.dataBrokerDataLock.RUnlock()
+
+	req := a.getEvaluatorRequestFromCheckRequest(in, sessionState)
+	reply, err := a.pe.Evaluate(ctx, req)
 	if err != nil {
+		log.Error().Err(err).Msg("error during OPA evaluation")
 		return nil, err
 	}
-	logAuthorizeCheck(ctx, in, reply, rawJWT)
+	logAuthorizeCheck(ctx, in, reply, sessionState)
 
 	switch {
-	case reply.GetHttpStatus().GetCode() > 0 && reply.GetHttpStatus().GetCode() != http.StatusOK:
-		// custom error from the IsAuthorized call
-		return a.deniedResponse(in,
-			reply.GetHttpStatus().GetCode(),
-			reply.GetHttpStatus().GetMessage(),
-			reply.GetHttpStatus().GetHeaders(),
-		), nil
-
-	case reply.Allow:
-		// ok!
-		return a.okResponse(reply, rawJWT, isNewSession), nil
-
-	case reply.SessionExpired,
-		errors.Is(sessionErr, sessions.ErrExpired),
-		errors.Is(sessionErr, sessions.ErrIssuedInTheFuture),
-		errors.Is(sessionErr, sessions.ErrMalformed),
-		errors.Is(sessionErr, sessions.ErrNoSessionFound),
-		errors.Is(sessionErr, sessions.ErrNotValidYet):
-		// redirect to login
-
-		// no redirect for forward auth, that's handled by a separate config setting
+	case reply.Status == http.StatusOK:
+		return a.okResponse(reply, rawJWT), nil
+	case reply.Status == http.StatusUnauthorized:
 		if isForwardAuth {
 			return a.deniedResponse(in, http.StatusUnauthorized, "Unauthenticated", nil), nil
 		}
-
 		return a.redirectResponse(in), nil
-
-	default:
-		// all other errors
-		var msg string
-		if sessionErr != nil {
-			msg = sessionErr.Error()
-		}
-		return a.deniedResponse(in, http.StatusForbidden, msg, nil), nil
 	}
+	return a.deniedResponse(in, int32(reply.Status), reply.Message, nil), nil
 }
 
-func (a *Authorize) getEnvoyRequestHeaders(rawJWT []byte, isNewSession bool) ([]*envoy_api_v2_core.HeaderValueOption, error) {
-	var hvos []*envoy_api_v2_core.HeaderValueOption
+func (a *Authorize) waitForSessionAndUser(timeout time.Duration, sessionID string) {
+	ctx, clearTimeout := context.WithTimeout(context.Background(), timeout)
+	defer clearTimeout()
 
-	if isNewSession {
-		cookieStore, err := getCookieStore(a.currentOptions.Load(), a.currentEncoder.Load())
-		if err != nil {
-			return nil, err
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	var userID string
+
+	for {
+		a.dataBrokerDataLock.RLock()
+		s, ok := a.dataBrokerData.Get(sessionTypeURL, sessionID).(*session.Session)
+		a.dataBrokerDataLock.RUnlock()
+
+		if ok {
+			if s.GetDeletedAt() != nil {
```

Comment:
This can be simplified to just:

```
if s.GetUserId() != "" {
    userID = s.GetUserId()
    break
}
return
```

Reply:
a session with a `DeletedAt` set will also have a `UserId` set, so we have to check `DeletedAt` first